pkg/julia/manifest/parse.go

```diff
@@ -0,0 +1,159 @@
+		return nil, nil, xerrors.Errorf("seek error: %w", err)
+	}
+
+	// naive parser to get line numbers
+	pkgParser := naivePkgParser{r: r}
+	lineNumIdx := pkgParser.parse()
+
+	var libs []types.Library
+	var deps []types.Dependency
+	for name, manifestDeps := range man.Dependencies {
+		for _, manifestDep := range manifestDeps {
+			version := depVersion(&manifestDep, man.JuliaVersion)
+			pkgID := manifestDep.UUID
+			lib := types.Library{
+				ID:      pkgID,
+				Name:    name,
+				Version: version,
+			}
+			if pos, ok := lineNumIdx[manifestDep.UUID]; ok {
+				lib.Locations = []types.Location{{StartLine: pos.start, EndLine: pos.end}}
+			}
+
+			libs = append(libs, lib)
+
+			if len(manifestDep.DependsOn) > 0 {
+				deps = append(deps, types.Dependency{
+					ID:        pkgID,
+					DependsOn: manifestDep.DependsOn,
+				})
+			}
+		}
+	}
+	sort.Sort(types.Libraries(libs))
+	sort.Sort(types.Dependencies(deps))
+	return libs, deps, nil
+}
+
+// Returns the effective version of the `dep`.
+// stdlib packages do not have a version in the manifest because they are packaged with julia itself
+func depVersion(dep *primitiveDependency, juliaVersion string) string {
+	if len(dep.Version) == 0 {
+		return juliaVersion
+	}
+	return dep.Version
+}
+
+// Decodes a primitive manifest using the metadata from parse time.
+func decodeManifest(man *primitiveManifest, metadata *toml.MetaData) (*primitiveManifest, error) {
+	// Decode each dependency into the new manifest
+	for depName, primDeps := range man.Dependencies {
+		var newPrimDeps []primitiveDependency
+		for _, primDep := range primDeps {
+			newPrimDep, err := decodeDependency(man, primDep, metadata)
+			if err != nil {
+				return nil, err
+			}
+			newPrimDeps = append(newPrimDeps, newPrimDep)
+		}
+		man.Dependencies[depName] = newPrimDeps
+	}
+
+	return man, nil
+}
+
+// Decodes a primitive dependency using the metadata from parse time.
+func decodeDependency(man *primitiveManifest, dep primitiveDependency, metadata *toml.MetaData) (primitiveDependency, error) {
+	// Try to decode as []string first where the manifest looks like deps = ["A", "B"]
+	var possibleDeps []string
+	err := metadata.PrimitiveDecode(dep.Dependencies, &possibleDeps)
+	if err == nil {
+		var possibleUuids []string
+		for _, depName := range possibleDeps {
+			primDep := man.Dependencies[depName]
+			if len(primDep) > 1 {
+				return primitiveDependency{}, xerrors.Errorf("Dependency %q has invalid format (parsed multiple deps): %s", depName, primDep)
+			}
+			possibleUuids = append(possibleUuids, primDep[0].UUID)
+		}
+		dep.DependsOn = possibleUuids
+		return dep, nil
+	}
+
+	// The other possibility is a map where the manifest looks like
+	// [deps.A.deps]
+	// B = "..."
+	var possibleDepsMap map[string]string
+	err = metadata.PrimitiveDecode(dep.Dependencies, &possibleDepsMap)
+	if err == nil {
+		possibleUuids := maps.Values(possibleDepsMap)
+		dep.DependsOn = possibleUuids
```

Comment:
If i remember correctly - `maps.Values` returns values in random order. Can you add sort here?

Reply:
Done